lib/rust/profiler/data/src/lib.rs

```diff
@@ -0,0 +1,686 @@
+//! Interface to profile data.
+//!
+//! # Overview
+//!
+//! Usage of this API starts with [`interpret`], which converts a raw log of [`profiler::Event`]s
+//! into a hierarchy of [`Measurement`] values. Each [`Measurement`] contains all information
+//! relating to a particular profiler. This conversion checks low-level invariants of the event log
+//! format, and produces an output that is easier to use than the raw events.
+//!
+//! # Usage example: storing and retrieving metadata
+//!
+//! ```
+//! use enso_profiler as profiler;
+//! use enso_profiler_data as profiler_data;
+//! use profiler::profile;
+//!
+//! // A metadata type.
+//! #[derive(serde::Serialize, serde::Deserialize, PartialEq, Eq, Debug)]
+//! struct MyData(u32);
+//!
+//! // An activity that produces metadata.
+//! struct ActivityWithMetadata {
+//!     meta_logger: profiler::MetadataLogger<MyData>,
+//!     // ...fields for doing stuff
+//! }
+//! impl ActivityWithMetadata {
+//!     fn new() -> Self {
+//!         let meta_logger = profiler::MetadataLogger::register("MyData");
+//!         Self { meta_logger /* ... */ }
+//!     }
+//!
+//!     #[profile(Objective)]
+//!     fn action_producing_metadata(&self) {
+//!         let x = MyData(23);
+//!         self.meta_logger.log(x);
+//!     }
+//! }
+//!
+//! // Run the activity that produces metadata, and profile it.
+//! #[profile(Objective)]
+//! fn demo() {
+//!     let act = ActivityWithMetadata::new();
+//!     act.action_producing_metadata();
+//! }
+//!
+//! fn store_and_retrieve_metadata() {
+//!     demo();
+//!
+//!     #[derive(serde::Deserialize, PartialEq, Eq, Debug)]
+//!     enum MyMetadata {
```

Comment:
I think this enum might require a bit more explanation. How is this matched to the serialized data? I assume internally this is matched on the JSON representation, but this is not obvious in this example.

Reply:
When I was looking for a good place to document this in the API, I realized `take_log<Metadata>`/`interpret` is a poor entry point to this API. I've changed it so `take_log` now returns the log as a JSON `String`, and `Measurement::parse` interprets a log. Then there's a place to document how the metadata enum is used: `Measurement`.

Also, I hadn't implemented error-recovery for metadata changes, so I added that and a `skip_failed_metadata` test case.